### PR TITLE
[MRG] Mondrian Tree intuition

### PR DESCRIPTION
**Pull request by MechCoder**
Fixes https://github.com/scikit-garden/scikit-garden/issues/1

### CODE REVIEW

File: docs/MondrianTreeRegressor_intuition.md [hunk 1]
```diff
@@ -0,0 +1,277 @@
...
+
+If you stare at the equation for half a minute, that makes sense as well because the larger the bounding box of a node, the time of split is smaller. Also the above property sets the time of split of leaves to infinity. Once the split feature and threshold are decided, then the tree construction happens in a similar way to that of a decision tree, a node splits the training data into two parts, $X[:, f] < \delta$ to one leaf and $X[:, f] > \delta$ to the other.    `
+
+###  Prediction time
+
+Recall that for a decision tree, for a new training point, computing the predictive mean and variance is fairly straightforward. That is, find the leaf node that a new training point lands in and output the mean and variance of the node.
+
+The prediction step of a Mondrian Tree is a bit more complicated. It takes into account all the nodes in the path of a new point from the root to the leaf for making a prediction. Mathematically, the distribution of $P(Y | X)$ is given by
+
+$P(Y | X) = \sum_{j} w_j \mathcal{N} (m_j, v_j)$
```
Comment (MechCoder) [inline docs/MondrianTreeRegressor_intuition.md]: The latex does not render well here. I'll fix it all at once.

**Comment by MechCoder** [comment]
cc: @glouppe @balajiln

File: docs/examples/MondrianTreeRegressor.md [hunk 2]
```diff
@@ -0,0 +1,273 @@
...
+import matplotlib.pyplot as plt
+from sklearn.tree import ExtraTreeRegressor
+from skgarden import MondrianTreeRegressor
+from itertools import cycle
+%matplotlib inline
+```
+
+##  Extremely Randomized Tree - Recap
+
+In a tree-based algorithm, the decision rule at every node is constructed by considering a set of candidate split points and that candidate split point that minimizes a certain impurity (generally the mse in case of continuous values) is chosen. In an extremely randomized tree, the number of candidate splits `S` are limited to `n_features` and each candidate split `S[i]` is drawn unifomly from the bounds `(l_f[i], u_f[i])`.
```
Comment (glouppe) [inline docs/examples/MondrianTreeRegressor.md]: they are limited to `max_features`

File: docs/examples/MondrianTreeRegressor.md [hunk 3]
```diff
@@ -0,0 +1,273 @@
...
+
+In a single extremely randomized tree, this construction might seem suboptimal but while constructing an ensemble of trees, it makes sure that each tree constructed in an independent fashion. In addition, limiting the number of candidate split points makes the tree construction very fast. Let us now generate some toy data.
```
Comment (glouppe) [inline docs/examples/MondrianTreeRegressor.md]: "it makes sure that each tree constructed in an independent fashion"

I would complete this argument by saying that decorrelating predictions in an ensemble is precisely the key factor to achieve lower generalization error.

File: docs/examples/MondrianTreeRegressor.md (hunk 2 above)
Comment (glouppe) [inline docs/examples/MondrianTreeRegressor.md]: also, i would rather say that we aim to "maximize the decrease of impurity in the resulting child nodes (generally the reduction of variance in case of continuous output values)"

File: docs/examples/MondrianTreeRegressor.md [hunk 4]
```diff
@@ -0,0 +1,273 @@
...
+plt.show()
+```
+
+
+![png](../mondrian_tree/plot1.png)
+
+
+## Plotting decision boundaries using ERT's
+
+Let us now scikit-learn's ExtraTreeRegressor to train on the generated toy data, predict on some unseen data and plot decision boundaries in the 1-D space. Also, we set the `max_depth` parameter to 2, which means there can be a maximum of 4 decision boundaries in the 1-D space.
```
Comment (glouppe) [inline docs/examples/MondrianTreeRegressor.md]: a verb is missing in the first sentence

File: docs/examples/MondrianTreeRegressor.md [hunk 5]
```diff
@@ -0,0 +1,273 @@
...
+We notice that as we move away from the training data
+
+1. The predicted mean remains constant.
+2. It is determined by the decision split at the bounds of the space.
+
+Ideally, as move away from the training data we are unsure about the target value to predict and would like to fall back on some prior mean. The mondrian tree solves this problem in a very intelligent way.
+
+##  Mondrian Tree
+
+### Train time
```
Comment (glouppe) [inline docs/examples/MondrianTreeRegressor.md]: I would rename this section -- it feels like it concerns timing issues when building a tree :)

File: docs/examples/MondrianTreeRegressor.md [hunk 6]
```diff
@@ -0,0 +1,273 @@
...
+
+At every node, the split threshold and feature is decided independently of the target or the decrease in impurity! Yes, that is right.
```
Comment (glouppe) [inline docs/examples/MondrianTreeRegressor.md]: Would be nice to note that this is almost equivalent to extra-trees built with `max_features=1`

File: docs/examples/MondrianTreeRegressor.md [hunk 7]
```diff
@@ -0,0 +1,273 @@
...
+
+1. The split feature index `f` is drawn with a probability proportional to `u_b[f] - l_b[f]` where `u_b` and `l_b` and the upper and lower bounds of all the features.
+2. After fixing the feature index, the split threshold \(\delta\) is then drawn from a uniform distribution with limits `l_b`, `u_b`.
+
+The intuition being that a feature that has a huge difference between the bounds is likelier to be an "important" feature.
+
+At every node, in a decision tree, the mean and impurity are stored (or class probabilities in case of classification). In addition to this, every node `j` in a mondrian tree tracks two other things.
+
+1. The upper and lower bounds of all the features in that particular node.
+2. The time of split, which is drawn from an exponential with mean \(\sum_{f=1}^D(u_b[f] - l_b[f])\)
```
Comment (glouppe) [inline docs/examples/MondrianTreeRegressor.md]: Not sure it is clear what "time of split" means

File: docs/examples/MondrianTreeRegressor.md [hunk 8]
```diff
@@ -0,0 +1,273 @@
...
+
+\(w_j(x) = 1 - \sum_{k \in anc(j)} w_k(x)\)
+
+Let us take a more than half a minute to stare at these equations and understand what they mean.
+
+1. \(p_j(x)\) is high when \(\eta_{j}(x)\) is high. As \(\eta_{j}(x)\) approaches infinity, \(p_j(x)\) approaches zero. This means that when the point is far away from the bounding box of the node, the probability of separation becomes high.
+2. \(p_j(x)\) is high when \(\Delta{j}\) is high. This means when the bounding box of a node is small as compared to the bounding box of its parent, the probability of separation becomes high.
+3. The weight given to each node, \(w_j\) can be decomposed into two factors, \(p_j\) and \(\prod_{k \in anc(j)} (1 - p_k(x))\). The product term can be understood as the probability of that point not being separated till that particular node is reached. So the node where a new point just starts to branch of is given the highest weight.
+4. For a point in the training data, \(p_j(x)\) becomes zero all nodes (and hence \(w_j(x)\)). The leaf then has a weightage of 1.0 and this reduces to a standard decision tree prediction.
+
```
Comment (glouppe) [inline docs/examples/MondrianTreeRegressor.md]: We should also explain what happens for test points far away from the training data.

File: docs/examples/MondrianTreeRegressor.md (hunk 5 above)
Comment (balajiln) [inline docs/examples/MondrianTreeRegressor.md]: can you add a link to my thesis? not many people might be familiar with mondrian forests. feel free to re-use any figures from my thesis if that's helpful :)

File: docs/examples/MondrianTreeRegressor.md (hunk 7 above)
Comment (balajiln) [inline docs/examples/MondrianTreeRegressor.md]: Figure 5.1 highlights difference between mondrian tree and vanilla decision tree.

Another way to think of "Time of split" is the following: imagine that edges between parent and child nodes are associated with a non-negative weight (we need a special parent(root) node). Then the time of split at a node is the sum of the weights along the path from parent(root) to the node. If weights are all 1, the time of split is the depth of the node. So we can think of time of split as "weighted depth".

File: docs/examples/MondrianTreeRegressor.md (hunk 5 above)
Comment (balajiln) [inline docs/examples/MondrianTreeRegressor.md]: agree about training time: "training mode" and "prediction mode" maybe?

**Comment by MechCoder** [comment]
I'm planning to move the predictive mean and variance topic to a separate discussion.

File: docs/examples/MondrianTreeRegressor.md (hunk 7 above)
Comment (MechCoder) [inline docs/examples/MondrianTreeRegressor.md]: Nice!

**Comment by MechCoder** [comment]
I spent some time rewriting most of it. You can find the rendered webpage over here. (https://65-88780543-gh.circle-artifacts.com/0/tmp/circle-artifacts.KYpGCOL/site/examples/MondrianTreeRegressor/index.html)

Please let me know how it looks.

File: docs/examples/MondrianTreeRegressor.md [hunk 9]
```diff
@@ -0,0 +1,302 @@
...
+from skgarden import MondrianTreeRegressor
+from itertools import cycle
+%matplotlib inline
+```
+
+##  A decision tree regressor
+
+A decision tree is one of the easier-to-understand machine learning algorithms. While training, the input training space `X` is recursively partitioned into a number of rectangular subspaces. While predicting the label of a new point, one determines the rectangular subspace that it falls into and outputs the label representative of that subspace. This is usually the mean of the labels for a regression problem.
+
+The rectangular subspaces are constructed in a greedy manner doing a binary partition at a time, hence each split is determined by a (f, \(\delta)\) tuple. The point `X` for which `X[f]` is lesser than \(\delta\) are in the left subspace and vice versa.
```
Comment (glouppe) [inline docs/examples/MondrianTreeRegressor.md]: small x instead?

File: docs/examples/MondrianTreeRegressor.md [hunk 10]
```diff
@@ -0,0 +1,302 @@
...
+
+## So how is each split done?
+
+The split tuple (f, \(\delta)\) is determined as that combination that creates child nodes such that the weighted decrease in impurity is maximum. Mathematically, this is the combination that maximises, `N_{parent} Imp_{Parent} - N_{left} Imp}_{left} - N_{right} Imp{right}`
```
Comment (glouppe) [inline docs/examples/MondrianTreeRegressor.md]: missing dot at the end of the sentence.

File: docs/examples/MondrianTreeRegressor.md [hunk 11]
```diff
@@ -0,0 +1,302 @@
...
+
+In a standard decision tree, this combination is found out by searching though all possible combinations of feature indices and values in the training data. That sounds pretty expensive!
+
+In an extremely randomized tree, this is made much faster by limiting the number of candidate splits `S` are limited to another hyperparameter `max_features`. Each candidate split `S[i]` is then drawn uniformly from the bounds `(l_f[i], u_f[i])`.
```
Comment (glouppe) [inline docs/examples/MondrianTreeRegressor.md]: The first sentence does not make sense.

Also it is not clear what you are referring to by a split `S`.

A split should be defined as a tuple (f, threshold) (as you define it above). `max_features` then controls the subset of features to consider for candidate splits (and not the number of candidate splits as currently phrased).

File: docs/examples/MondrianTreeRegressor.md [hunk 12]
```diff
@@ -0,0 +1,302 @@
...
+plt.show()
+```
+
+
+![png](../mondrian_tree/plot1.png)
+
+
+## Plotting decision boundaries using ERT's
+
+Let us now use scikit-learn's ExtraTreeRegressor to train on the generated toy data, predict on some unseen data and plot decision boundaries in the 1-D space. Also, we set the `max_depth` parameter to 2, which means there can be a maximum of 4 decision boundaries in the 1-D space.
```
Comment (glouppe) [inline docs/examples/MondrianTreeRegressor.md]: can you put backticks around `ExtraTreeRegressor`?

File: docs/examples/MondrianTreeRegressor.md [hunk 13]
```diff
@@ -0,0 +1,302 @@
...
+
+A decision tree is one of the easier-to-understand machine learning algorithms. While training, the input training space `X` is recursively partitioned into a number of rectangular subspaces. While predicting the label of a new point, one determines the rectangular subspace that it falls into and outputs the label representative of that subspace. This is usually the mean of the labels for a regression problem.
+
+The rectangular subspaces are constructed in a greedy manner doing a binary partition at a time, hence each split is determined by a (f, \(\delta)\) tuple. The point `X` for which `X[f]` is lesser than \(\delta\) are in the left subspace and vice versa.
+
+## So how is each split done?
+
+The split tuple (f, \(\delta)\) is determined as that combination that creates child nodes such that the weighted decrease in impurity is maximum. Mathematically, this is the combination that maximises, `N_{parent} Imp_{Parent} - N_{left} Imp}_{left} - N_{right} Imp{right}`
+
+In a standard decision tree, this combination is found out by searching though all possible combinations of feature indices and values in the training data. That sounds pretty expensive!
```
Comment (glouppe) [inline docs/examples/MondrianTreeRegressor.md]: Importantly, this search is guided by a maximization of impurity decrease, which takes into account how better we would become for predicting the output.

I think this is important to underline, as mondrian trees skip this step.

**Comment by MechCoder** [comment]
Addressed.

**Comment by MechCoder** [comment]
I've reworded it a bit more.

**Comment by glouppe** [comment]
Thanks! On my side, this looks good enough for a first version. 

What do you think @balajiln ?

**Comment by MechCoder** [comment]
All right, merging for now. @balajiln Please do have a look when you get the time. I can address your comments even after it has been merged.

**Comment by MechCoder** [comment]
Example is now at https://scikit-garden.github.io/examples/MondrianTreeRegressor/

**Comment by balajiln** [comment]
Hi Manoj, Gilles, 

Sorry for the delay. I reviewed the version in https://scikit-garden.github.io/examples/MondrianTreeRegressor/

Overall, it looks really good :) Just have a few minor comments:

Plotting decision boundaries using ERT's: Would be useful to add a colorbar to the figure as the target values corresponding to predictions are not clear.

"Every subspace j in a mondrian tree also stores information about." Use ":" at end instead of "." as you're talking about 1 and 2 below. I thought the sentence was incomplete when I first read it ...

I'd suggest adding a remark in the beginning something like "the distribution of Mondrian trees trained on a dataset in an incremental fashion is the same as that of Mondrian trees trained on the same dataset in a batch fashion, irrespective of the order in which the data points are observed" as none of the online random forests have this property. This property also motivates why Mondrian tree splits without looking at labels and uses the particular weighting for branching off etc, otherwise it sounds a bit like a hack :) You don't need to discuss the online version,  you can say you're going to focus only on the uncertainty aspect for this tutorial.

Hope that helps!

Cheers,
Balaji
